### PR TITLE
feat: add seal command for kubeseal

### PR DESCRIPTION
**Pull request by paddatrapper**
### Description

Adds support for generating sealedsecrets using kubeseal

### Breaking Change

- [ ] Yes
- [x] No

### Testing Notes

**What I did:**
Replaced `kubeseal` with `karina seal` in calls to create sealedsecrets

### **Links**

Fixes: #645

### CODE REVIEW

File: cmd/seal.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+		if err := kubeseal(flagString + argString); err != nil {
+			log.Fatalf("failed to run kubeseal: %v", err)
+		}
+	},
+}
+
+func init() {
+	Seal.Flags().String("cert", "", "Certificate/public key file/URL to use for encryption")
+	Seal.Flags().String("controller-namespace", metav1.NamespaceSystem, "Namespace of sealed-secrets-controller")
+	Seal.Flags().String("controller-name", "sealed-secrets-controller", "Name of sealed-secrets controller.")
```
Comment (moshloop) [inline cmd/seal.go:99]: So these fields should be based on platform.SealedSecrets, the interface for the user would be:

`karina seal secret.yaml -c karina.yaml > sealed.yaml`
and 
`karina unseal sealed.yaml -c karina.yaml > secret.yaml`

File: cmd/unseal.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+			os.Remove(privateKeyFile)
+			log.Fatalf("failed to run kubeseal: %v", err)
+		}
+		os.Remove(privateKeyFile)
+	},
+}
+
+func init() {
+	Unseal.Flags().StringP("format", "o", "json", "Output format for sealed secret. Either json or yaml")
+	Unseal.Flags().Bool("allow-empty-data", false, "Allow empty data in the secret object")
```
Comment (moshloop) [inline cmd/unseal.go:128]: I think we can remove this option

File: cmd/unseal.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+		if err := kubeseal(flagString + argString); err != nil {
+			os.Remove(privateKeyFile)
+			log.Fatalf("failed to run kubeseal: %v", err)
+		}
+		os.Remove(privateKeyFile)
+	},
+}
+
+func init() {
+	Unseal.Flags().StringP("format", "o", "json", "Output format for sealed secret. Either json or yaml")
```
Comment (moshloop) [inline cmd/unseal.go:127]:
```suggestion
	Unseal.Flags().StringP("format", "o", "yaml", "Output format for sealed secret. Either json or yaml")
```

File: cmd/seal.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+		flagString := strings.Join(flags, " ")
+		argString := strings.Join(args, " ")
+		if err := kubeseal(flagString + argString); err != nil {
+			log.Fatalf("failed to run kubeseal: %v", err)
+		}
+	},
+}
+
+func init() {
+	Seal.Flags().StringP("format", "o", "json", "Output format for sealed secret. Either json or yaml")
```
Comment (moshloop) [inline cmd/seal.go:93]: If you move the declaration of Seal into the init function, then you can define the variables up front, and then use StringVar, BoolBar etc without the need for getFlag functions

Comment (paddatrapper) [reply cmd/seal.go:93]: Those getFlag functions also add the required `--name` flags that are passed to the kubeseal function call. They would still be needed for the filepath and slice values, but could be removed for the others

Comment (paddatrapper) [reply cmd/seal.go:93]: Moving the declaration also doesn't make the getFlagBool calls any simpler, as they need to conditionally add a flag based on the input.

File: cmd/unseal.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+			log.Fatalf("Unable to write sealed-secrets-key to disk: %v", err)
+		}
+		kubeseal := p.GetBinary("kubeseal")
+		flagString := strings.Join(flags, " ")
+		argString := strings.Join(args, " ")
+		if err := kubeseal(flagString + argString); err != nil {
+			os.Remove(privateKeyFile)
+			log.Fatalf("failed to run kubeseal: %v", err)
+		}
+		os.Remove(privateKeyFile)
```
Comment (moshloop) [inline cmd/unseal.go:122]:
```suggestion
		err = ioutil.WriteFile(privateKeyFile, file, 0600)
		defer 		os.Remove(privateKeyFile)
		if err != nil {
			log.Fatalf("Unable to write sealed-secrets-key to disk: %v", err)
		}
		kubeseal := p.GetBinary("kubeseal")
		flagString := strings.Join(flags, " ")
		argString := strings.Join(args, " ")
		if err := kubeseal(flagString + argString); err != nil {
			log.Fatalf("failed to run kubeseal: %v", err)
		}
```

File: cmd/unseal.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+			log.Fatalf("Unable to resolve file path for %s: %v", privateKeyFile, err)
+		}
+		flags := []string{
+			"--recovery-unseal",
+			getFlagString("format", cmd),
+			getFlagBool("allow-empty-data", cmd),
+			fmt.Sprintf("--recovery-private-key %s", filePath),
+		}
+		if !p.SealedSecrets.Disabled && p.SealedSecrets.Certificate != nil {
+			if p.SealedSecrets.Certificate.Cert != "" {
```
Comment (moshloop) [inline cmd/unseal.go:87]: If we dont't have a copy of the cert shouldn't we fail fast ?

File: cmd/seal.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+		argString := strings.Join(args, " ")
+		if err := kubeseal(flagString + argString); err != nil {
+			log.Fatalf("failed to run kubeseal: %v", err)
+		}
+	},
+}
+
+func init() {
+	Seal.Flags().StringP("format", "o", "yaml", "Output format for sealed secret. Either json or yaml")
+	Seal.Flags().StringP("sealed-secret-file", "w", "", "Sealed-secret (output) file")
```
Comment (moshloop) [inline cmd/seal.go:94]: This shouldn't be necessary - we always output to stdout

File: cmd/seal.go [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+	Seal.Flags().Bool("fetch-cert", false, "Write certificate to stdout. Useful for later use with --cert")
```
Comment (moshloop) [inline cmd/seal.go:95]: Why do we need this ?

File: cmd/unseal.go [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+	Run: func(cmd *cobra.Command, args []string) {
+		p := getPlatform(cmd)
+		filePath, err := filepath.Abs(privateKeyFile)
+		if err != nil {
+			log.Fatalf("Unable to resolve file path for %s: %v", privateKeyFile, err)
+		}
+		flags := []string{
+			"--recovery-unseal",
+			getFlagString("format", cmd),
+			getFlagBool("allow-empty-data", cmd),
```
Comment (moshloop) [inline cmd/unseal.go:83]: Why do we need this ?

File: cmd/seal.go [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+	Seal.Flags().Bool("validate", false, "Validate that the sealed secret can be decrypted")
+	Seal.Flags().String("merge-into", "", "Merge items from secret into an existing sealed secret file, updating the file in-place instead of writing to stdout.")
+	Seal.Flags().Bool("raw", false, "Encrypt a raw value passed via the --from-* flags instead of the whole secret object")
```
[the remark this answers is not in the document]
Comment (paddatrapper) [reply cmd/seal.go:98]: merge-into was helpful several times during the BCB setup when we needed to append a secret to an existing sealed-secret. The others I'll remove

**Pull request merged**